storage/queue.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package storage provides implementations and shared components for tessera storage backends.
+package storage
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/globocom/go-buffer"
+)
+
+// Queue knows how to queue up a number of entries in order, taking care of deduplication as they're added.
+//
+// When the buffered queue grows past a defined size, or the age of the oldest entry in the
+// queue reaches a defined threshold, the queue will call a provided FlushFunc with
+// a slice containing all queued entries in the same order as they were added.
+//
+// If multiple identical entries are added to the queue between flushes, the queue will deduplicate them by
+// passing only the first through to the FlushFunc, and returning the index assigned to that entry to all
+// duplicate add calls.
+// Note that this deduplication only applies to "in-flight" entries currently in the queue; entries added
+// after a flush will not be deduped against those added before the flush.
+type Queue struct {
+	buf   *buffer.Buffer
+	flush FlushFunc
+
+	inFlightMu sync.Mutex
+	inFlight   map[string]*entry
+}
+
+// IndexFunc is a function which returns an assigned log index, or an error.
+type IndexFunc func() (idx uint64, err error)
+
+// FlushFunc is the signature of a function which will receive the slice of queued entries.
+// It should return the index assigned to the first entry in the provided slice.
+type FlushFunc func(ctx context.Context, entries [][]byte) (index uint64, err error)
```

Comment:
This assumes that all the entries in a queue will be sequenced one after other one. If the flush function does some deduplication, this property might not hold. Could you return an index for each entry instead please?

Reply:
That is a requirement for using this mix-in, yup.

If there's a reason why a given storage implementation cannot meet this requirement then it's fine for it to do its own queueing, but we're explicitly targeting dedup as "best effort" for performance reasons, so I'd expect it should be fine (unless there's a storage implementation which physically _cannot_ store dupes?)

Returning an index for each entry is something I'd rather avoid if possible (what would the API look like - return a `map[string([]byte)]uint64` and force the caller to iterate over it and reconcile what's happened to the missing entries?)

Reply:
You could return a map, or []uint64, that is the same length that entries is. It still allows deduplication as best effort, but has the merit of pushing all the sequencing logic to the flush function rather than then computing the index for each entry after Flush has returned.

Reply:
I think we should wait until there's actually a concrete storage implementation which _requires_ this before making that change - this is an internal component and we're nowhere near a tagged release yet so it'll be easy to make.

Reply:
Al and I had a long conversation over VC about this.

Summary:
I ran into scaling problems because of deduplication twice already, both with DynamoDB and Aurora. We need to do more tests to see how expensive, scalable and fast it is to do deduplication using S3/GCS for CT deployments. In the interest of moving forward let's keep deduplication in S3/GCS for now, even if we know that it will have a hit on latency, which means that batch deduplication is not a thing for now, and come back to this later. I have both S3 and Aurora deduplication experimental code ready, I'll try and run both with the same QPS to see what's the cost difference over a few hours.